File: NewsApp/app/src/main/java/com/example/android/topic41/presentation/viewmodel/ArticlesViewModel.java
package com.example.android.topic41.presentation.viewmodel;

import android.util.Log;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.example.android.topic41.domain.usecase.SingleUseCaseInterface;
import com.example.android.topic41.domain.util.Article;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;
import io.reactivex.Observer;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;


public class ArticlesViewModel extends ViewModel {
    private SingleUseCaseInterface useCase;
    private boolean isFirstRequest = true;
    private boolean isInitializedActivity = false;

    private MutableLiveData<List<Article>> temp = new MutableLiveData<>();
    private MutableLiveData<String> errorMessage = new MutableLiveData<>();
    private MutableLiveData<Boolean> isLoadingState = new MutableLiveData<>(false);
    private MutableLiveData<Boolean> isShowedErrorMessage = new MutableLiveData<>(true);

    public ArticlesViewModel(SingleUseCaseInterface useCase) {
        Log.i("mLog_VIEWMODEL", "I am new again.");
        this.useCase = useCase;
    }

    public void loadArticles(String theme) {
        if(isFirstRequest || isInitializedActivity) {
            isLoadingState.setValue(true);
            useCase.loadArticles(theme)
                    .toObservable()
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(new Observer<List<Article>>() {
                        @Override
                        public void onSubscribe(Disposable d) {
                        }

                        @Override
                        public void onNext(List<Article> articles) {
                            setArticles(articles);
                        }

                        @Override
                        public void onError(Throwable e) {
                            setErrorMessage(e.getMessage());
                        }

                        @Override
                        public void onComplete() {
                        }
                    });
            isFirstRequest = false;
        }
        isInitializedActivity = true;
    }

    public void errorMessageShowed() {
        isShowedErrorMessage.setValue(true);
    }

    public void activityDestroyed() {
        isInitializedActivity = false;
    }

    public LiveData<List<Article>> getArticles() {
        return temp;
    }
    public LiveData<Boolean> getIsLoadingState() {
        return isLoadingState;
    }
    public LiveData<Boolean> getIsShownErrorMessage() {
        return isShowedErrorMessage;
    }
    public LiveData<String> getErrorMessage() {
        return errorMessage;
    }


    public void setArticles(List<Article> articles) {
        temp.setValue(articles);
        isLoadingState.setValue(false);
    }


    public void setErrorMessage(String message) {
        errorMessage.setValue(message);
        temp.setValue(new ArrayList<>());
        isShowedErrorMessage.setValue(false);
        isLoadingState.setValue(false);
    }
}

File: NewsApp/app/src/main/java/com/example/android/topic41/presentation/viewmodel/ArticlesViewModel.kt
package com.example.android.topic41.presentation.viewmodel

import android.annotation.SuppressLint
import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.example.android.topic41.domain.usecase.SingleUseCaseInterface
import com.example.android.topic41.domain.util.Article
import io.reactivex.android.schedulers.AndroidSchedulers
import java.util.*


class ArticlesViewModel(useCase: SingleUseCaseInterface) : ViewModel() {
    private val useCase: SingleUseCaseInterface = useCase
    private var isFirstRequest = true
    private var isInitializedActivity = false
    private val temp = MutableLiveData<List<Article>>()
    private val errorMessage = MutableLiveData<String>()
    private val isLoadingState = MutableLiveData(false)
    private val isShowedErrorMessage = MutableLiveData(true)

    @SuppressLint("CheckResult")
    fun loadArticles(theme: String) {
        if (isFirstRequest || isInitializedActivity) {
            isLoadingState.value = true
            useCase.loadArticles(theme)
                    .toObservable()
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe({articles: List<Article> ->setArticles(articles) },
                        { e: Throwable -> e.message?.let { setErrorMessage(it) } })
            isFirstRequest = false
        }
        isInitializedActivity = true
    }

    fun errorMessageShowed() {
        isShowedErrorMessage.value = true
    }

    fun activityDestroyed() {
        isInitializedActivity = false
    }

    val articles: LiveData<List<Article>>
        get() = temp

    fun getIsLoadingState(): LiveData<Boolean> {
        return isLoadingState
    }

    val isShownErrorMessage: LiveData<Boolean>
        get() = isShowedErrorMessage

    fun getErrorMessage(): LiveData<String> {
        return errorMessage
    }

    private fun setArticles(articles: List<Article>) {
        temp.value = articles
        isLoadingState.value = false
    }

    private fun setErrorMessage(message: String) {
        errorMessage.value = message
        temp.value = ArrayList()
        isShowedErrorMessage.value = false
        isLoadingState.value = false
    }
}


